From db2b3f0d0633d0f78f3f3167ad488520f49cc32b Mon Sep 17 00:00:00 2001
From: Nikolajs Agafonovs <nikolajs.agafonovs@gmail.com>
Date: Wed, 21 Oct 2026 16:20:08 +0300
Subject: [PATCH] tccgen: GCC cast-to-union extension

A cast to a union type with a member of the value's type yields a
temporary object of the union type (an lvalue) with that member set.
Arrays and functions decay before being matched; incomplete unions, a
missing member and use outside a function are errors.
---
 tccgen.c | 38 +++++++++++++++++++++++++++++++++++++-
 1 file changed, 37 insertions(+), 1 deletion(-)

diff --git a/tccgen.c b/tccgen.c
index 104f0b8..4b0cd80 100644
--- a/tccgen.c
+++ b/tccgen.c
@@ -3622,6 +3622,39 @@ static void cast_error(CType *st, CType *dt)
     type_incompatibility_error(st, dt, "cannot convert '%s' to '%s'");
 }
 
+/* GCC extension: cast to a union type having a member of the type of
+   the value. The result is a temporary union object (an lvalue). */
+static void gen_cast_to_union(CType *type)
+{
+    Sym *f;
+    int size, align, addr;
+    CType st = vtop->type;
+
+    if (type->ref->c < 0)
+        tcc_error("cast to incomplete union type");
+    /* arrays and functions decay before being matched with a member */
+    st.t &= ~(VT_ARRAY | VT_VLA);
+    if ((st.t & VT_BTYPE) == VT_FUNC)
+        mk_pointer(&st);
+    for (f = type->ref->next; f; f = f->next)
+        if (!(f->type.t & VT_BITFIELD)
+            && is_compatible_unqualified_types(&f->type, &st))
+            break;
+    if (!f)
+        cast_error(&vtop->type, type);
+    if (!local_scope)
+        tcc_error("cast to union type outside of a function");
+    size = type_size(type, &align);
+    loc = (loc - size) & -align;
+    addr = loc;
+    /* store the value into the member, then push the union object */
+    vset(&f->type, VT_LOCAL | VT_LVAL, addr + f->c);
+    vswap();
+    vstore();
+    vpop();
+    vset(type, VT_LOCAL | VT_LVAL, addr);
+}
+
 /* verify type compatibility to store vtop in 'dt' type */
 static void verify_assign_cast(CType *dt)
 {
@@ -5780,7 +5813,10 @@ ST_FUNC void unary(void)
                 return;
             } else {
                 unary();
-                gen_cast(&type);
+                if (IS_UNION(type.t) && (vtop->type.t & VT_BTYPE) != VT_STRUCT)
+                    gen_cast_to_union(&type);
+                else
+                    gen_cast(&type);
             }
         } else if (tok == '{') {
 	    int saved_nocode_wanted = nocode_wanted;
